src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs

```diff
@@ -0,0 +1,191 @@
+            // at Microsoft.Spark.Sql.PicklingUdfWrapper`2.Execute(Int32 splitIndex, Object[] input, Int32[] argOffsets) in Microsoft.Spark\Sql\PicklingUdfWrapper.cs:line 44
+            // at Microsoft.Spark.Worker.Command.PicklingSqlCommandExecutor.SingleCommandRunner.Run(Int32 splitId, Object input) in Microsoft.Spark.Worker\Command\SqlCommandExecutor.cs:line 239
+            // at Microsoft.Spark.Worker.Command.PicklingSqlCommandExecutor.ExecuteCore(Stream inputStream, Stream outputStream, SqlCommand[] commands) in Microsoft.Spark.Worker\Command\SqlCommandExecutor.cs:line 139
+            Func<Column, Column> udf = Udf<IDictionary<string, string>, string>(
+                    dict => dict.Count.ToString());
+
+            DataFrame df = _df.WithColumn("tempName", Map(_df["name"], _df["name"]));
+            Assert.Throws<Exception>(() => df.Select(udf(df["tempName"])).Show());
+
+            // Currently, there is a workaround to support MapType using Hashtable. See the example below.
+            Func<Column, Column> workingUdf = Udf<Hashtable, string>(
+                dict => dict.Count.ToString());
+
+            Row[] rows = df.Select(workingUdf(df["tempName"])).Collect().ToArray();
+            Assert.Equal(3, rows.Length);
+
+            var expected = new[] { "1", "1", "1" };
+            for (int i = 0; i < rows.Length; ++i)
+            {
+                Row row = rows[i];
+                Assert.Equal(1, row.Size());
+                Assert.Equal(expected[i], row.GetAs<string>(0));
+            }
+        }
+
+        /// <summary>
+        /// UDF that returns Map type.
+        /// </summary>
+        [Fact]
+        public void TestUdfWithReturnAsMapType()
+        {
+            // UDF with return as MapType throws a following exception:
+            // Unhandled Exception: System.Reflection.TargetInvocationException: Exception has been thrown by the target of an invocation. 
+            // ---> System.NotImplementedException: The method or operation is not implemented.
+            // System.NotImplementedException: The method or operation is not implemented.
+            // at Microsoft.Spark.Sql.Row.Convert() in Microsoft.Spark\Sql\Row.cs:line 169
+            // at Microsoft.Spark.Sql.Row..ctor(Object[] values, StructType schema) in Microsoft.Spark\Sql\Row.cs:line 34
+            // at Microsoft.Spark.Sql.RowConstructor.GetRow() in MicrTosoft.Spark\Sql\RowConstructor.cs:line 113
+            // at Microsoft.Spark.Sql.RowCollector.Collect(ISocketWrapper socket) + MoveNext() in Microsoft.Spark\Sql\RowCollector.cs:line 36
+            // at Microsoft.Spark.Sql.DataFrame.GetRows(String funcName) + MoveNext() in Microsoft.Spark\Sql\DataFrame.cs:line 891
+            Func<Column, Column> udf = Udf<string, IDictionary<string, string>>(
+                str => new Dictionary<string, string> { { str, str } });
+            Assert.Throws<NotImplementedException>(() => _df.Select(udf(_df["name"])).Collect().ToArray());

```

Comment:
over 100 char limit

Reply:
Done for all over 100 char limit. Thanks!